Add -add-k8s-namespace-suffix flag to the sync-catalog command

ishustava opened the pull request:
This PR proposes to add `-add-k8s-namespace-suffix` flag to the `sync-catalog` command to resolve #111 and fix #139:
* When true, sync catalog process will append k8s namespace
  of services it's syncing to Consul, separated by an underscore.
  E.g for a service `foo` in the `default` namespace, it will register
  a Consul service called `foo_default`. The underscore was chosen as
  a delimiter, as opposed to a dash, to avoid name collisions. When
  using a dash as a delimiter, both service `foo` in the namespace `bar-baz`
  and service `foo-bar` in the namespace `baz` would be registered as
  `foo-bar-baz`. Because underscore is not a valid character for a
  resource name in k8s, but it is in Consul, it's safe to use as a
  delimiter.
* False by default to allow users to opt in. This also allows us
  to keep the default behavior for now that is currently documented
  [here](https://www.consul.io/docs/platform/k8s/service-sync.html#service-name)
* Add tests for the sync-catalog command and separate common test helper
  functions into a shared package.

adilyse commented on subcommand/sync-catalog/command_test.go:
```diff
@@ -0,0 +1,210 @@
+}
+
+// This function starts the command asynchronously and returns a blocking chan.
+// When finished, the command will send its exit code to the channel.
```
What happens if the command doesn't terminate? Could that cause the tests that use this to run indefinitely?

ishustava replied:
That's a great question. Since this function is for running sync-catalog command which is long-running, it's the responsibility of the caller to terminate it by calling `cmd.interrupt`. I've added a comment clarifying that.

adilyse commented on subcommand/sync-catalog/command_test.go:
```diff
@@ -0,0 +1,206 @@
+	exitChan := runCommandAsynchronously(&cmd, []string{
+		"-http-addr", testAgent.HTTPAddr(),
+	})
+
```
```suggestion
```

adilyse commented on subcommand/sync-catalog/command_test.go:
```diff
@@ -0,0 +1,206 @@
+		"-consul-write-interval", "500ms",
+		"-add-k8s-namespace-suffix",
+	})
+
```
```suggestion
```

adilyse commented on subcommand/sync-catalog/command_test.go:
```diff
@@ -0,0 +1,206 @@
+		"-consul-write-interval", "1s",
+		"-add-k8s-namespace-suffix",
+	})
+
```
```suggestion
```

adilyse commented on subcommand/sync-catalog/command_test.go:
```diff
@@ -0,0 +1,206 @@
+		"-consul-write-interval", "1s",
+		"-add-k8s-namespace-suffix",
+	})
+
```
```suggestion
```

lkysow commented on helper/testutil/k8s.go:
```diff
@@ -0,0 +1,30 @@
+)
+
+// LBService returns a Kubernetes service of type LoadBalancer.
+func LBService(name, lbIP string) *apiv1.Service {
```
I'm not a big fan of having this in a shared package. I'd prefer to see it duplicated in each test package because it's easier to modify if a specific package needs to change its LB and it's not a lot of code.

ishustava replied:
Ok, yeah I think that seems to be the coding preference at HashiCorp. I've removed this helper and made it more "WET" 😁

ishustava commented on catalog/to-consul/resource.go:
```diff
@@ -659,9 +665,21 @@ func (t *serviceEndpointsResource) Delete(key string) error {
+		// 'foo-bar-baz'. Because underscore is not a valid character for a
+		// resource name in k8s, but it is in Consul, it's safe to use as a
+		// delimiter.
+		return fmt.Sprintf("%s_%s", name, svc.Namespace)
```
Agree! Thank you for catching that. Updated.

ishustava commented on catalog/to-consul/resource.go:
```diff
@@ -71,6 +71,12 @@ type ServiceResource struct {
+	// AddK8SNamespaceSuffix set to true appends Kubernetes namespace
+	// to the service name being synced to Consul separated by a an underscore.
+	// For example, service 'foo' in the default namespace will be synced
+	// as 'foo_default'. Default is false.
```
Sounds good. Updated.

ishustava commented on catalog/to-consul/resource_test.go:
```diff
@@ -242,6 +243,101 @@ func TestServiceResource_changeSyncToFalse(t *testing.T) {
+
+	// Insert an LB service with the sync=true
+	svc := testutil.LBService("foo", "1.2.3.4")
+	_, err := client.CoreV1().Services("bar").Create(svc)
```
Done!

ishustava commented:
@lkysow and @adilyse thank you both for the thoughtful reviews! I've updated the delimiter to a dash and incorporated your suggestions. 

The main reason for this decision is that underscores are not valid in DNS names. DNS interface feels like an important feature if using Consul for service discovery, and breaking it doesn't feel like the right direction. Although, as mentioned in the description, collisions are still possible with a dash, this solution provides sufficient protection against the most common/obvious naming collisions.

ishustava commented on catalog/to-consul/resource_test.go:
```diff
@@ -1261,7 +1356,7 @@ func lbService(name string) *apiv1.Service {
 			LoadBalancer: apiv1.LoadBalancerStatus{
 				Ingress: []apiv1.LoadBalancerIngress{
 					{
-						IP: "1.2.3.4",
```
It is not necessary, that's true. I do think though that it makes the test more readable because when you assert on IP a few lines below you don't have to drill down to the function to see what IP it's setting there.

ishustava commented on catalog/to-consul/resource.go:
```diff
@@ -659,9 +665,16 @@ func (t *serviceEndpointsResource) Delete(key string) error {
 }
 
-func (t *ServiceResource) prefixServiceName(name string) string {
+func (t *ServiceResource) addPrefixAndK8SNamespace(svc *apiv1.Service) string {
```
Changed the signature to `func (t ServiceResource) addPrefixAndK8SNamespace(name, namespace string) string`.

filintod commented:
An opinion, I do like this workaround but the real solution is for Consul to support namespaces. Connect seems to support it.  This is set to be set as unplanned goal of Consul 1.7 (https://github.com/hashicorp/consul/milestone/45).  How will this work with that solution?

adilyse commented:
@filintod Indeed, namespaces will be added as an Enterprise feature to Consul in the near future and will be supported in Kubernetes. This is still a useful fix until that happens and for folks not running the Enterprise version.

adilyse commented on catalog/to-consul/resource_test.go:
```diff
@@ -1245,8 +1340,8 @@ func TestServiceResource_clusterIPTargetPortNamed(t *testing.T) {
 
-// lbService returns a Kubernetes service of type LoadBalancer.
-func lbService(name string) *apiv1.Service {
+// LBService returns a Kubernetes service of type LoadBalancer.
```
```suggestion
// lbService returns a Kubernetes service of type LoadBalancer.
```

adilyse commented on subcommand/sync-catalog/command_test.go:
```diff
@@ -0,0 +1,228 @@
+	}
+}
+
+// LBService returns a Kubernetes service of type LoadBalancer.
```
```suggestion
// lbService returns a Kubernetes service of type LoadBalancer.
```
